rfd/0019-event-iteration-api.md

```diff
@@ -0,0 +1,114 @@
+
+### GRPC Endpoints 3.2
+
+This RFD calls for two new endpoints. One endpoint for out-of-session event requests and one for in-session event requests.
+
+These endpoints will live on the `AuthService` GRPC service in [api/client/proto/authservice.proto](https://github.com/gravitational/teleport/blob/master/api/client/proto/authservice.proto) and are defined as follows:
+
+```proto
+rpc GetEvents(GetEventsRequest) returns (Events);
+```
+
+```proto
+rpc GetSessionEvents(GetSessionEventsRequest) returns (Events);
+```
+
+### Implementation details 3.3
+
+#### Auth service 3.3.1
+
+Handling of these new GRPC endpoints will be implemented in [lib/auth/grpcserver.go](https://github.com/gravitational/teleport/blob/master/lib/auth/grpcserver.go). `auth.Client.SearchEvents` and `auth.Client.SearchSessionEvents` will be refactored with pagination support implemented with the same startkey/lastkey interface as seen in the GRPC API. These methods are what the GRPC handlers call. The current HTTP API implementation will be slightly altered to deal with these changes but will remain identicatal in functionality, simlpy not exposing the new internal pagination support.
+
+#### Backends 3.3.2
+
+Performed some investigative work on how we can implement this pagination support in the different backends. The standard solution that works with the current backend interface is to simply query all events in a given date range from the backend and then filter out the paginated section using the given start key and limit. This is however something we want to avoid as it involves a lot of uneeded processing on every request by the authentication server.
+
+Pagination support will needed to be added to the `IAuditLog` interface defined in `lib/events/api.go`. This involves adding a start key function parameter and an optional last key parameter in the return value of the `SearchEvents` and `SearchSessionEvents` methods.
+
+Pagination in DynamoDB seems possible to implement using start keys (partition + sort key) and global indexes for requesting only the paginated events and should be compatible with our current format.
+
+Not an expert on the format used for filelog but since it is a plain sequence of JSON documents, implementing efficient pagination seems difficult. On this backend we might need to fall back to loading all events into memory and then sorting out the events that fall within the range of events requested.
+
+Firestore seems to have good support for pagination with it's query cursors and document snapshots which allow us to define a range between startkey and startkey + limit. This means we can efficiently query a subset of events.
+
+As for the etcd backend, the official Go client now has support for API v3 which has good pagination support with the `RangeRequest` message and initial investigation shows that implementing pagination in our backend using may require some refactoring but is realistic.
```

Comment:
We don't store events in etcd, so this is not necessary

Reply:
Thank, will remove and add a section I missed to commit on sqlite support.

Reply:
👍 We don't appear to store events in sqlite either do we? Only dynamodb and firestore right?

Reply:
I think the other supported backends for events are `stdout` and `file://` (possibly also `syslog`)

Reply:
You can't query events from `stdout` and `syslog` can you? How would that work?

Reply:
Sorry, perhaps 'backend' was the wrong term here. We support writing events to all the listed destinations under `audit_events_uri` in the [config reference](https://goteleport.com/teleport/docs/config-reference/), which are `dynamodb`, `firestore`, `file` and `stdout`. Indeed, Teleport uses `file` out of the box as the default so presumably has some kind of rudimentary pagination implementation over that which should be included.

I've no idea what happens if you set the only destination for `audit_events_uri` as `stdout` - I presume Teleport refuses to start, as the web UI wouldn't be able to query it.